Supports separate encoders with vanilla RNNs

kylebgorman opened the pull request:
This PR adds support for separate encoders in vanilla RNNs; it addresses #313 (but doesn't touch transformers so it does not close it). We adopt the merging approach from the hard attention models, which works with both "linear" and RNN features encoders. Suppose that the source encoder generates $B \times L \times E$ and the features encoder generates $B \times L' \times E'$. To create the final encoding, we first average features encodings across all positions producing a tensor of size $B \times 1 \times E'$. We then expand this tensor to $B \times L \times E'$ by broadcasting it across all positions in the source tensor. We then concatenate along the encoding dimension, producing a new tensor of size $B \times L \times (E + E')$.  The effect can be likened to spreading the features encoder information across all positions and thus roughly doubling the size of the encoding. 

# Fixes

* The `LinearEncoder` layer is misnamed because it's just a non-contextual embedding, with no actual affine projections. It is renamed `EmbeddingEncoder` and documentation is adjusted to match. 
* A separate `attention_input_size` argument is no longer necessary because it is always the same as `decoder_input_size`.
* The `NotImplementedError` is removed because it is in fact implemented.

# Alternatives considered

I experimented with concatenating along the length dimension instead. This works when there are two RNNs of the same hidden size and dimensionality but would break if the features encoder is just an embedding because E and E' may not match (indeed, one common rule of thumb is to use a hidden layer size 4x of the embedding size). We _could_ also add an affine $E' \times E$ projection linear layer to the embedding encoder, but that'd only be applicable for vanilla RNNs and wouldn't make sense elsewhere. If one wants a length concatenation effect, one can probably just use a single source/features encoder and the default concatenation, and the proposal in #301 provides an even more principled generalization of that idea. So I reject this idea.

# Future work

Complete #313 by adding something appropriate for transformers.

Adamits commented:
Sorry for my delay.

Before I look at code, just want to make sure I understand:

> We then concatenate along the encoding dimension, producing a new tensor of size B×L×(E+E′).

Here L is the sequence length right (I was reading it as Layer at first)?  and L' was the seq length of the features? So we average over the feature representations, and concatenate that averaged vector onto each source symbol hidden representation.

Then for the alternate consideration:
> ...experimented with concatenating along the length dimension instead.
I am sort of confused about what this means. Are you just concatenating the matrices of symbol and feature encodings so we have a single sequence again (though the encodings happen by separate modules here ofc)?

I will remark that a third possibility is that you do no extra steps when encoding, and the decoder must attend to both the symbols and the features separately and then combine them (concatenation, affine transformer). I think this is what the pointer generator did at first. However, this complicates Yoyodyne a bit since then the decoder has to know if a separate features encoder was requested, and then conditionally attend to both and combine.

kylebgorman commented:
> Here L is the sequence length right (I was reading it as Layer at first)? and L' was the seq length of the features? So we average over the feature representations, and concatenate that averaged vector onto each source symbol hidden representation.

Correct.

>> ...experimented with concatenating along the length dimension instead.
> I am sort of confused about what this means. Are you just concatenating the matrices of symbol and feature encodings so we have a single sequence again (though the encodings happen by separate modules here ofc)?

Yes, that is the alternative I considered.


Sorry for my delay.

Before I look at code, just want to make sure I understand:

We then concatenate along the encoding dimension, producing a new tensor of size B×L×(E+E′).

Here L is the sequence length right (I was reading it as Layer at first)? and L' was the seq length of the features? So we average over the feature representations, and concatenate that averaged vector onto each source symbol hidden representation.

Then for the alternate consideration:

...experimented with concatenating along the length dimension instead.
I am sort of confused about what this means. Are you just concatenating the matrices of symbol and feature encodings so we have a single sequence again (though the encodings happen by separate modules here ofc)?

> I will remark that a third possibility is that you do no extra steps when encoding, and the decoder must attend to both the symbols and the features separately and then combine them (concatenation, affine transformer). I think this is what the pointer generator did at first. However, this complicates Yoyodyne a bit since then the decoder has to know if a separate features encoder was requested, and then conditionally attend to both and combine.

This is just the second possibility for me: it would be necessary to add a linear layer to what I'm now calling the `EmbeddingEncoder` to make it work in general because you don't expect the embedding dimensionality and hidden layer dimensionality to match. I don't think it's much of a complication for vanilla RNNs (it can be handled in the embedding encoder module plus the shared `forward` method for the abstract RNN model class and just in a few lines), I just think it's unlikely to be meaningfully different than using a single encoder anyways. 

It would be a potential complication for transformers with their fixed attention context though.

Adamits commented:
> This is just the second possibility for me: it would be necessary to add a linear layer to what I'm now calling the `EmbeddingEncoder` to make it work in general because you don't expect the embedding dimensionality and hidden layer dimensionality to match. I don't think it's much of a complication for vanilla RNNs (it can be handled in the embedding encoder module plus the shared `forward` method for the abstract RNN model class and just in a few lines), I just think it's unlikely to be meaningfully different than using a single encoder anyways.

I think our attention module already does a projection though, no? Anyway, we probably want to replace that with a simpler KQV attention at some point..

> 
> It would be a potential complication for transformers with their fixed attention context though.

I am having trouble understanding why this complicates the transformer. Is that in the case that we have two separate encodings to attend to, since the cross-attention is built into the torch implementation of a transformer decoder already? I think for the transformer ptr-gen I had to subclass it (just realized that should probably be updated to match the LSTM ptr-gen changes).

Adamits reviewed (approved):
Code changes LGTM. I will remark that I spent time thinking through the semantics of the changes (the comments describing what is happening, and linear --> embedding) and agree with all of them.

Adamits commented:
One more though: does this generalize for the edit action transducer, and the pointer generator? I think those might have had their own mechanisms for handling features, and I wonder if this conflicts with them?

kylebgorman commented:
> I think our attention module already does a projection though, no? Anyway, we probably want to replace that with a simpler KQV attention at some point..

I don't see how that's relevant. Attention takes a source tensor as an argument, and you logically can't concatenate tensors B x L x E (source encoding from an RNN encoder) and B x L' x E' (features encoding from the misnamed "linear encoder") on the second dimension if E != E', as will usually be the case under the current design.

So the alternative design requires me to add a linear projection of shape E' x E to the so-called linear encoder. It would look up the embedding and then apply the linear projection. (Right now it just looks up the embedding.) This would only be applicable for vanilla RNNs (it wouldn't make sense for pointer-generators or hard monotonic models, which use different strategies) and my proposal is that we don't bother.

> I am having trouble understanding why this complicates the transformer.

The transformer is implemented with a max source length for the positional embedding, and the RNNs don't. We could exceed this limit if we're doing additional concatenation along the length dimension. For instance if the max length is 128 and we have a long source string of 125 symbols and there are 5 features it'll crash (probably in a nasty way without an informative error message).

kylebgorman commented:
> One more though: does this generalize for the edit action transducer, and the pointer generator? I think those might have had their own mechanisms for handling features, and I wonder if this conflicts with them?

It matches the choice made in the hard monotonic model and the transducer. [It is different from the pointer-generator, which has a separate features attention module](https://github.com/CUNY-CL/yoyodyne/blob/e43920ff30affd4b71e084627eecf77e33fcec4e/yoyodyne/models/pointer_generator.py#L229). The transformer doesn't support this approach yet and I haven't studied the problem there.

So for me there are two other possibilities at most:

* the "alternate solution" described in the OP
* the pointer-generator solution linked to in this post

Adamits commented:
> > One more though: does this generalize for the edit action transducer, and the pointer generator? I think those might have had their own mechanisms for handling features, and I wonder if this conflicts with them?
> 
> It matches the choice made in the hard monotonic model and the transducer. [It is different from the pointer-generator, which has a separate features attention module](https://github.com/CUNY-CL/yoyodyne/blob/e43920ff30affd4b71e084627eecf77e33fcec4e/yoyodyne/models/pointer_generator.py#L229). The transformer doesn't support this approach yet and I haven't studied the problem there.
> 
> So for me there are two other possibilities at most:
> 
> * the "alternate solution" described in the OP
> * the pointer-generator solution linked to in this post

"The transformer doesn't support this approach"  --> I am fairly certain the pointer-generator solution is the one I implemented for the pointer generator transformer to be clear.

> The transformer is implemented with a max source length for the positional embedding

Oh so you are talking about the self-attention portion?

kylebgorman commented:
> "The transformer doesn't support this approach" --> I am fairly certain the pointer-generator solution is the one I implemented for the pointer generator transformer to be clear.

Oh, I was referring to the "vanilla transformer" here. [The pointer-generator transformer doesn't support separate features](https://github.com/CUNY-CL/yoyodyne/blob/e43920ff30affd4b71e084627eecf77e33fcec4e/yoyodyne/models/pointer_generator.py#L542) either though. (I added that check but that's because the features were being ignored otherwise.) 
 
> Oh so you are talking about the self-attention portion?

I wasn't making a distinction because I thought both had a length limit but I may be mistaken.

Adamits commented:
> > One more though: does this generalize for the edit action transducer, and the pointer generator? I think those might have had their own mechanisms for handling features, and I wonder if this conflicts with them?
> 
> It matches the choice made in the hard monotonic model and the transducer. [It is different from the pointer-generator, which has a separate features attention module](https://github.com/CUNY-CL/yoyodyne/blob/e43920ff30affd4b71e084627eecf77e33fcec4e/yoyodyne/models/pointer_generator.py#L229). The transformer doesn't support this approach yet and I haven't studied the problem there.

Ah ok. So I guess it supports separate features (https://github.com/CUNY-CL/yoyodyne/blob/e43920ff30affd4b71e084627eecf77e33fcec4e/yoyodyne/models/modules/transformer.py#L558C7-L558C32) but not a separate encoder.

> So for me there are two other possibilities at most:
> 
> * the "alternate solution" described in the OP
> * the pointer-generator solution linked to in this post


Ok, we can continue discussing for future work I guess. I think this PR looks good to go -- your solution implemented here seems like a good one.
